dist/js/select2.js

```diff
@@ -3306,11 +3306,11 @@ S2.define('select2/data/ajax',[
     }, this.ajaxOptions);
 
     if (typeof options.url === 'function') {
-      options.url = options.url(params);
+      options.url = options.url(self.$element, params);
     }
 
     if (typeof options.data === 'function') {
-      options.data = options.data(params);
+      options.data = options.data(self.$element, params);
```

Comment:
This should be `options.data.call(self.$element, params)` so the `this` is set instead of the first parameter.